# polyAudit
All-in-one poly(A) tail analyses for long-read RNA sequencing data

## Documentation

The entire pipeline is contained within `polya.py` and can be run in a single command. Default options will run the `-polyA` pipeline on the input file and write results to the output directory. Test input files can be found in `/test`.

    usage: polya.py [-h] [-p PRIMER_FILE] [-polyA] [-l LENGTH_FILE] [-k K] [-n N]
                    [-PAS PAS_FILE] [-pssm PSSM]
                    input_file output_directory

    Measure poly(A) tails, count k-mers, find the putative PAS, and generate PSSMs a given sequence file.

    positional arguments:
      input_file            A FASTA file with long-read sequencing data that hasnt
                            had poly(A) tails trimmed.
      output_directory      A path to the output directory.

    optional arguments:
      -h, --help            show this help message and exit
      -p PRIMER_FILE, --primer_file PRIMER_FILE
                            A file with a single primer sequences per line. NOTE:
                            it is recommended that the input sequences are trimmed
                            for adapters and primers prior to running this
                            program. I do not guarantee a desirable or predictable
                            outcome with the -p option.
      -polyA, --polyA       Runs the algorithm to measure poly(A) tails in a FASTA
                            file. Requires -p if sequences contain a 3' adapter or
                            primer. Will include poly(A) tail measurement, k-mer
                            counting in the 50 nt upstream of of the tail,
                            identification of a putative PAS, and generation of a
                            PSSM for the entire input file.
      -l LENGTH_FILE, --length_file LENGTH_FILE
                            A file with an isoform ID and poly(A) tail length on
                            each line to be used for k-mer counting and PAS
                            identification. This file can be generated with
                            -polyA.
      -k K, --k K           An integer, the kmer length (defaults to 6).
      -n N, --n N           An integer, the number of k-mers to pipe from the
                            k-mer search to the PAS search (defaults to 10).
      -PAS PAS_FILE, --pas_file PAS_FILE
                            Identify the most likely PAS; requires a file with one
                            k-mer per line. NOTE: the algorithm will search for
                            each k-mer one at a time and stop the search once it
                            has found one. Thus, it is HIGHLY recommended that
                            this file is ordered in descending PAS/k-mer usage.
      -pssm PSSM, --pssm PSSM
                            Generate a position-specific score matrix for the
                            region flanking the PAS. Requires a CSV file
                            (generated by -polyA) with column names Isoform, 3'
                            Sequence, kmer, and Count.

## Expected Output

### Default output files:

Command: `python polya.py test/sample_trimmed.fasta test/ -polyA`

  `sample_polya.csv`: A CSV file (without column names) that includes FASTA ID's from `input_file` in column one and the measured poly(A) tail length in column two.

  `sample_6mer.csv`: A CSV containing all possible k-mers of length `-k` with the measured abundance in the 50 nt upstream of the poly(A) tails in all FASTA entries in `input_file`.

  `sample_PAS.csv`: A CSV file with the predicted PAS for each FASTA entry in `input_file`. Also includes the index (i.e distance of the PAS upstream of the poly(A) tail).

  `sample_PAS_clean.fasta`: A FASTA file of all entries from `input_file` clipped to include the 15 nt downstream and 50 nt upstream of the 5' end of the poly(A) site.

  `sample_pssm.txt`: A position-specific score matrix (PSSM) generated from `sample_PAS_clean.fasta`.

### Potential additional files:

Command: `python polya.py test/sample.fasta test/ -p test/primers.txt -polyA`

  `sample_trimmed.fasta`: A FASTA files with sequences that have had 3' adapters/primers removed. NOTE: There are better tools to perform this trimming, and I recommend using those.

### Alternative command:

  A PSSM for the entire `input_file` is not that useful, so I recommend choosing subsets of sequences for PSSM generation (i.e. all transcripts that utilize a given PAS). In this case, you can bypass most functions and go directly to `-pssm`, for example:

  `python polya.py test/sample.fasta test/ -pssm test/AATAAA.csv`

  will result in:

  `AATAAA_pssm.txt` and `AATAAA_PAS_clean.fasta`

## Expected Data

I've also included an example R script for data analysis. Note that this script is NOT for deployment and should only be used as a guide. Following are potential plots that can be created using the data generated by this script.

![poly(A) Plot](/images/polyA.png)

(A) Histogram of poly(A) tail lengths with the median length drawn in red  
(B) Top ten 6-mers in the 50 nt upstream of the 5' end of the poly(A)  
(C) Location of given 6-mers in the 50 nt upstream of the 5' end of the poly(A)  
(D) Nucleotide frequencies in the 50 nt upstream of the 5' end of the poly(A), faceted by PAS  

## Conda Environment

    biopython                 1.74             py36h1de35cc_0
    blas                      1.0                    openblas
    bz2file                   0.98                     py36_1
    ca-certificates           2020.1.1                      0
    certifi                   2019.11.28               py36_0
    khmer                     3.0.0a2          py36h0a44026_1    bioconda
    libblas                   3.8.0               14_openblas    conda-    forge
    libcblas                  3.8.0               14_openblas    conda-    forge
    libcxx                    4.0.1                hcfea43d_1
    libcxxabi                 4.0.1                hcfea43d_1
    libedit                   3.1.20181209         hb402a30_0
    libffi                    3.2.1                h475c297_4
    libgfortran               4.0.0                         2    conda-    forge
    liblapack                 3.8.0               14_openblas    conda-    forge
    libopenblas               0.3.7                h3d69b6c_7    conda-    forge
    llvm-openmp               9.0.1                h28b9765_2    conda-    forge
    ncurses                   6.1                  h0a44026_1
    numpy                     1.13.1             py36_nomkl_0
    openssl                   1.1.1d               h1de35cc_4
    pandas                    0.24.2           py36h0a44026_0
    pip                       20.0.2                   py36_1
    python                    3.6.10               h359304d_0
    python-dateutil           2.8.1                      py_0
    pytz                      2019.3                     py_0
    readline                  7.0                  h1de35cc_5
    screed                    1.0.4                      py_0    bioconda
    setuptools                45.2.0                   py36_0
    six                       1.14.0                   py36_0
    sqlite                    3.31.1               ha441bb4_0
    tk                        8.6.8                ha441bb4_0
    wheel                     0.34.2                   py36_0
    xz                        5.2.4                h1de35cc_4
    zlib                      1.2.11               h1de35cc_3
